core/src/main/scala/tech/beshu/ror/accesscontrol/blocks/ImpersonationWarning.scala

```diff
@@ -0,0 +1,183 @@
+/*
+ *    This file is part of ReadonlyREST.
+ *
+ *    ReadonlyREST is free software: you can redistribute it and/or modify
+ *    it under the terms of the GNU General Public License as published by
+ *    the Free Software Foundation, either version 3 of the License, or
+ *    (at your option) any later version.
+ *
+ *    ReadonlyREST is distributed in the hope that it will be useful,
+ *    but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *    GNU General Public License for more details.
+ *
+ *    You should have received a copy of the GNU General Public License
+ *    along with ReadonlyREST.  If not, see http://www.gnu.org/licenses/
+ */
+package tech.beshu.ror.accesscontrol.blocks
+
+import cats.implicits.catsSyntaxOptionId
+import eu.timepit.refined.auto._
+import eu.timepit.refined.types.string.NonEmptyString
+import tech.beshu.ror.RequestId
+import tech.beshu.ror.accesscontrol.blocks.definitions.ldap.LdapService
+import tech.beshu.ror.accesscontrol.blocks.rules.AuthKeyHashingRule.HashedCredentials
+import tech.beshu.ror.accesscontrol.blocks.rules._
+import tech.beshu.ror.accesscontrol.blocks.rules.base._
+import tech.beshu.ror.accesscontrol.blocks.rules.base.impersonation.Impersonation
+import tech.beshu.ror.accesscontrol.blocks.rules.indicesrule.IndicesRule
+
+final case class ImpersonationWarning(block: Block.Name,
+                                      ruleName: Rule.Name,
+                                      message: NonEmptyString,
+                                      hint: String)
+
+object ImpersonationWarning {
+  sealed trait ImpersonationWarningSupport[+T <: Rule]
+
+  object ImpersonationWarningSupport {
+    trait WithPossibleWarnings[T <: Rule] extends ImpersonationWarningSupport[T] {
+      def warningFor(rule: T, blockName: Block.Name)
+                    (implicit requestId: RequestId): Option[ImpersonationWarning]
+    }
+
+    object WithPossibleWarnings {
+      def apply[T <: Rule](f: (T, Block.Name, RequestId) => Option[ImpersonationWarning]): WithPossibleWarnings[T] =
+        new WithPossibleWarnings[T] {
+          override def warningFor(rule: T, blockName: Block.Name)
+                                 (implicit requestId: RequestId): Option[ImpersonationWarning] = f(rule, blockName, requestId)
```

Comment:
do we need a `requestId` here?

Reply:
The `requestId` is passed to `tech.beshu.ror.accesscontrol.blocks.mocks.MocksProvider`